### ISSUE TITLE
reuse of an GPUAdapter to request a GPUDevice does not reject.

### ISSUE THREAD

**Issue by greggman** [gpuweb/gpuweb#4352]
TL;DR: Is there a reason why calling `requestDevice` on an invalid `GPUAdapter` doesn't reject? Current it resolves with a lost device.

---

I'm probably missing something but I think I'm not sure I understand this example.

In the spec [here](https://gpuweb.github.io/gpuweb/#initialization-examples) is this example

```js
let gpuDevice = null;

async function initializeWebGPU() {
    // Check to ensure the user agent supports WebGPU.
    if (!('gpu' in navigator)) {
        console.error("User agent doesn’t support WebGPU.");
        return false;
    }

    // Request an adapter.
    const gpuAdapter = await navigator.gpu.requestAdapter();

    // requestAdapter may resolve with null if no suitable adapters are found.
    if (!gpuAdapter) {
        console.error('No WebGPU adapters found.');
        return false;
    }

    // Request a device.
    // Note that the promise will reject if invalid options are passed to the optional
    // dictionary. To avoid the promise rejecting always check any features and limits
    // against the adapters features and limits prior to calling requestDevice().
    gpuDevice = await gpuAdapter.requestDevice();

    // requestDevice will never return null, but if a valid device request can’t be
    // fulfilled for some reason it may resolve to a device which has already been lost.
    // Additionally, devices can be lost at any time after creation for a variety of reasons
    // (ie: browser resource management, driver updates), so it’s a good idea to always
    // handle lost devices gracefully.
    gpuDevice.lost.then((info) => {
        console.error(`WebGPU device was lost: ${info.message}`);

        gpuDevice = null;

        // Many causes for lost devices are transient, so applications should try getting a
        // new device once a previous one has been lost unless the loss was caused by the
        // application intentionally destroying the device. Note that any WebGPU resources
        // created with the previous device (buffers, textures, etc) will need to be
        // re-created with the new one.
        if (info.reason != 'destroyed') {
            initializeWebGPU();
        }
    });

    onWebGPUInitialized();

    return true;
}

function onWebGPUInitialized() {
    // Begin creating WebGPU resources here...
}

initializeWebGPU();
```

A `GPUAdapter` is only allowed to be used one time to call `requestDevice`. Every other call to `requestDevice` will return a lost device. So, if you make a mistake and reuse the adapter, you'll end up in an infinite loop

```js
if (!('gpu' in navigator)) {
    console.error("User agent doesn’t support WebGPU.");
    return false;
}

const gpuAdapter = await navigator.gpu.requestAdapter();
if (!gpuAdapter) {
    console.error('No WebGPU adapters found.');
    return false;
}

let gpuDevice = null;

async function initializeWebGPU() {
    gpuDevice = await gpuAdapter.requestDevice();
    gpuDevice.lost.then((info) => {
        gpuDevice = null;
        if (info.reason != 'destroyed') {
            initializeWebGPU();
        }
    });

    onWebGPUInitialized();
    return true;
}

function onWebGPUInitialized() {
 ...
}

initializeWebGPU();
```

Because 99.9999% of the time the code above works (because you get a working device on the first call to `requestDevice`), you have no idea you're doing it wrong. When a user in the wild hits the lost device case, the page will just spin, with no indication anything is wrong and no errors?

It seems like `requestAdapter` should reject if the adapter is invalid?

**Comment by teoxoy** [comment]
Should this be M0 since it would be a backwards incompatible change?

**Comment by kdashg** [comment]
Mozilla thinks that we should differentiate the "sorry we failed" case of e.g. underlying Adapter loss with the "you violated the only-one-created-device" rule, where we would prefer the latter to Reject as an "api misuse error".

---

I do wonder if we should do something more aggressive to encourage authors to assume that adapters are flakey, like specing them to *start* expired 10% of the time.

I also think it would be cleaner to change the one-requestDevice-success-per-adapter rule to one-requestDevice-attempt-per-adapter. I think it would be encourage more robust code if the workflow for requesting a device *always* required requesting an adapter.

In some ways, this turns an Adapter into a one-shot DeviceRequestConfigurator in a two-step device creation workflow.

**Comment by Kangz** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/GPU-Web-2024-07-24#m0-reuse-of-an-gpuadapter-to-request-a-gpudevice-does-not-reject--issue-4352">GPU Web WG 2024-07-24 Atlantic-time</a></summary>

* **(consensus to change adapter reuse to a rejection)**
* CW: make things more brittle on purpose?
* KG: Moz's position - better to differentiate "sorry we failed" with adapter loss. "you already created one device". you're actively misusing the API.
* KN: the reason it's this way is we want one error handling path. If it doesn't reject immediately, it never will. However - you're expecting both to happen at the same time (requestAdapter + requestDevice). Failure could be a programming error. But I agree with the logic that this should be a rejection.
* CW: changes spec behavior - gives you an already-lost device? (Yes.) Side-effect is that it removes one of the encouragements for people to handle device loss. Want people to handle this case, though.
* KN: Don't think it's worth the pain of doing this. It's rare and people can reload. Also this should be extremely rare.
* RC: An adapter being lost can mean that it is lost forever, e.g. it was physically unplugged. In the Surface computer there's an Intel GPU in the tablet, and an Nvidia GPU in the keyboard. If you unplug them, the Nvidia GPU doesn't ever come back and adapters should be re-enumerated.
* JB: can also happen when someone undocks. Seems common. Think it's important to do this. wgpu probably isn't doing this right, either. We don't create a new factory on this event.
* KG: FF in browser does support this. Right now, GPUAdapter - once you create a device, you can't use it again.
* CW: the web has easy reload functionality - upon reload, will enumerate adapters again. Probably fine if the page breaks, Ctrl-R to reload?
* KG: we don't want to design APIs to be fragile. Rather, be robust. WebGPU apps don't always save state well enough to smoothly reload.
* MM: which proposal are we discussing?
* CW: we lost one way to encourage developers to handle device loss. Basically, do we need more encouragement in other ways. 
* KG: I don't think they're tied together. Have consensus to reject on API misuse. Other thing - we'd reject even if the adapter's lost.
* CW: sounds good.
* MM: APIs like Fetch reject upon network error. For similarity - even if website's correct but something happens in-browser, should still reject.
* KG: that requires people to handle rejection for both requestAdapter and requestDevice. Would like to pretend the adapter's fine.
* CW: because under the as-if rule, the adapter could be lost between the time you request the device, and get it back. Have to handle losing device before you get it.
* MM: returning the device that's lost could be handled in promise rejection. But I won't fight for it.
* KN: there are good reasons we don't do that, but maybe we don't need to discuss them live.

</details>
